sql/idxrecommendations: TestIndexRecommendationsStats failed

sql/idxrecommendations.TestIndexRecommendationsStats [failed](https://mesolite.cluster.engflow.com/invocations/default/70fe88c1-b37c-4402-b77a-df15e2f94f30?testReportRun=3&testReportShard=1&testReportAttempt=1#targets-Ly9wa2cvc3FsL2lkeHJlY29tbWVuZGF0aW9uczppZHhyZWNvbW1lbmRhdGlvbnNfdGVzdA==) on master @ [152c95d20a52357c31375d9464f24d54a7482a94](https://github.com/cockroachdb/cockroach/commits/152c95d20a52357c31375d9464f24d54a7482a94):


```
=== RUN   TestIndexRecommendationsStats
    test_log_scope.go:172: test logs captured to: outputs.zip/logTestIndexRecommendationsStats3173708565
    test_log_scope.go:82: use -show-logs to present logs inline
    test_server_shim.go:168: automatically injected an external process virtual cluster under test; see comment at top of test_server_shim.go for details.
    idx_recommendations_test.go:104: -- test log scope end --
test logs left over in: outputs.zip/logTestIndexRecommendationsStats3173708565
--- FAIL: TestIndexRecommendationsStats (240.16s)
=== RUN   TestIndexRecommendationsStats/index_recommendations_generated
    idx_recommendations_test.go:69: condition failed to evaluate within 3m45s: from testutils.go:401: expected 1 statement(s), got 2 filters=query=SELECT * FROM t WHERE v > _,exec_count=7
        
        goroutine dump: /mnt/engflow/worker/work/0/exec/bazel-out/k8-fastbuild/testlogs/pkg/sql/idxrecommendations/idxrecommendations_test/run_3_of_3/test.outputs/goroutine_dump_2430547113.txt
--- FAIL: TestIndexRecommendationsStats/index_recommendations_generated (232.73s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>deadlock=true</code>
 - <code>run=3</code>
 - <code>shard=1</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestIndexRecommendationsStats.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-63823

--- a/pkg/sql/idxrecommendations/BUILD.bazel
+++ b/pkg/sql/idxrecommendations/BUILD.bazel
@@ -46,6 +46,7 @@ go_test(
         "//pkg/testutils/testcluster",
         "//pkg/util/leaktest",
         "//pkg/util/log",
+        "//pkg/util/timeutil",
         "@com_github_cockroachdb_datadriven//:datadriven",
         "@com_github_stretchr_testify//require",
     ],

--- a/pkg/sql/idxrecommendations/idx_recommendations_test.go
+++ b/pkg/sql/idxrecommendations/idx_recommendations_test.go
@@ -8,15 +8,18 @@ package idxrecommendations_test
 import (
 	"context"
 	"testing"
+	"time"
 
 	"github.com/cockroachdb/cockroach/pkg/base"
 	"github.com/cockroachdb/cockroach/pkg/sql/idxrecommendations"
 	"github.com/cockroachdb/cockroach/pkg/sql/opt/indexrec"
+	"github.com/cockroachdb/cockroach/pkg/sql/sqlstats"
 	"github.com/cockroachdb/cockroach/pkg/sql/sqlstats/persistedsqlstats/sqlstatstestutil"
 	"github.com/cockroachdb/cockroach/pkg/testutils/serverutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/sqlutils"
 	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
 	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
 	"github.com/stretchr/testify/require"
 )
 
@@ -25,7 +28,20 @@ func TestIndexRecommendationsStats(t *testing.T) {
 	defer log.Scope(t).Close(t)
 
 	ctx := context.Background()
-	testServer, sqlConn, _ := serverutils.StartServer(t, base.TestServerArgs{})
+
+	// Pin all executions to a single SQL stats aggregation window. Without
+	// this, a test run that crosses an hour boundary produces duplicate
+	// (fingerprint, aggregatedTs) entries, which both inflates the count
+	// asserted by WaitForStatementEntriesEqual and makes the index
+	// recommendation visible in only one of the two buckets that QueryRow
+	// then reads non-deterministically.
+	stubTime := timeutil.Now().Truncate(time.Hour)
+	sqlStatsKnobs := sqlstats.CreateTestingKnobs()
+	sqlStatsKnobs.StubTimeNow = func() time.Time { return stubTime }
+
+	testServer, sqlConn, _ := serverutils.StartServer(t, base.TestServerArgs{
+		Knobs: base.TestingKnobs{SQLStatsKnobs: sqlStatsKnobs},
+	})
 	defer testServer.Stopper().Stop(ctx)
 
 	testConn := sqlutils.MakeSQLRunner(sqlConn)

